### ISSUE TITLE
Manage log group created when enabling API Gateway execution logs

### ISSUE THREAD

**Issue by coyoteecd** [serverless/serverless#8882]
### Use case description

Serverless Framework allows you to enable detailed execution logs at API Gateway level. This is described in [docs here](https://www.serverless.com/framework/docs/providers/aws/events/apigateway#logs). When execution logs are enabled, AWS will automatically create a log group in CloudWatch named `API_Gateway_Execution_Log_[unique-id]/[stage-name]`.

Currently, this has two limitations:
* the `logRetentionInDays` property is not applied to the log group. Likely because the log group is not part of the CloudFormation template, also not created by Serverless explicitly. See https://github.com/serverless/serverless/pull/6548#issuecomment-557086071
* when removing the stage, the log group is left behind, even though the API gateway is removed

Full execution data contains HTTP request/response body, so is not something you want to accumulate in a production environment. AWS recommends that they should be disabled, however they may be useful to investigate rarely occurring issues that are hard to reproduce. I do not want to rely on manual actions to set the log retention, nor do additional cleanup after a stack is removed.

### Proposed solution

I see that the access log group is already removed through the AWS SDK [when updating a stage](https://github.com/serverless/serverless/blob/61dd3bde8d17cdd995fdd27259a689d12bee1e42/lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.js#L379). I think a similar approach could work. 

Therefore, when a stage is created/updated, we would check if execution logs are enabled and if so, use describeLogGroups to see if the log group is there (can use the hard-coded `API-Gateway-Execution-Logs-` prefix as input, then iterate the result looking for the stage we're deploying. If the group is there, use putRetentionPolicy with the global setting configured in provider.

When a stage is removed (or updated with the execution logs disabled), we would again look for the log group for the stage we're removing, then use deleteLogGroup to get rid of it.

### Note

I considered implementing the above in a plugin; however I think it also makes sense having it as a core functionality. Opened this to gather opinions, if it makes sense I'm willing to contribute a PR provided I'm given some directions of where to put the code.

@medikoo?

**Comment by pgrzesik** [comment]
Thanks @coyoteecd - I've noticed similar thing about execution logs and I believe it would be great to address the mentioned issues. One thing I'm not sure about is using the current approach with SDK calls as it's very error prone and has it's own issues at the moment: https://github.com/serverless/serverless/issues/6906
I think it would be better to first migrate it to either pure CloudFormation (if there's support for that) or to Custom Resource as suggested in https://github.com/serverless/serverless/issues/7600. 

Please let me know what do you think. @medikoo your input will also be much appreciated as I believe you know the issues with current setup quite well

**Comment by coyoteecd** [comment]
CloudFormation is a no-go I think, since AWS does not have support for it.
Custom Resources: I only ran into these once (also with setting up logs and the shared cloudwatch role ARN) and the UX was poor, but maybe I was just unlucky. At any rate, my knowledge on the internals of serverless package is lacking and wouldn't see myself doing that move.

Meanwhile, since I need this in our project, I chose to solve the problem locally via a plugin that implements the functionality described above (see https://github.com/coyoteecd/serverless-api-gateway-execution-log-manager). Not perfect, since it probably misses corner-cases such as those mentioned above, but sufficient as a stop-gap solution.

**Comment by medikoo** [comment]
@coyoteecd I think what might have work (@pgrzesik we've discussed that today) is to configure `API_Gateway_Execution_Log_[unique-id]/[stage-name]` log group in CF stack whenever those logs for API Gateway are turned on, as I assume having it there will ensure automatic removal of this log group (with it's content) once the setting is gone from the configuration (and in result from CF stack)

It's actually how it works with lambdas (it's IAM log write access rights that actually creates and write logs and not defining them in a stack. We define them, only to be able to remove them when given lambdas or stack is removed)

**Comment by coyoteecd** [comment]
I've thought about that direction, however I have no idea how it would work, since the log group is created (by API Gateway) during stack deployment using a generated "unique id" in the name. It's a chicken-and-egg problem, I have to deploy the stack to figure out the name, by which time the stack template is already in S3.

With lambdas it works because you can tell AWS to write logs to a specific ARN. But I didn't find a way to configure it through CloudFormation. As far as I see the settings are defined here: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-apitgateway-stage-methodsetting.html; you can enable traces, but there's no ARN to specify. Am I missing something?

As a side-topic, I wonder why Serverless is setting up access logs and tracing levels via API Gateway "updateStage" calls, given that they are configurable via CloudFormation? Maybe this part should indeed be moved to the generated stack template.

**Comment by medikoo** [comment]
> I've thought about that direction, however I have no idea how it would work, since the log group is created (by API Gateway) during stack deployment using a generated "unique id" in the name.

In that case indeed it might be impossible. Are you sure this unique id part doesn't reference e.g. API id? Because if that's the case, then I think we would be able to reference name correctly with help of CF intrinsic functions

> As a side-topic, I wonder why Serverless is setting up access logs and tracing levels via API Gateway "updateStage" calls, given that they are configurable via CloudFormation? Maybe this part should indeed be moved to the generated stack template.

At first we were providing that via CloudFormation, but then we realized those settings works only when API Gateway is created and any updates to it are ineffective (PR that changed that: https://github.com/serverless/serverless/pull/6084)

Of course it might have changed since then, so it might be good to confirm weather it's still the case

**Comment by coyoteecd** [comment]
> Are you sure this unique id part doesn't reference e.g. API id?

Sigh, you're right, it is the API ID.
So in this case we could have serverless "generate" a LogGroup entry in the template and hope that API Gateway won't complain when it's trying to do the same, correct?

**Comment by medikoo** [comment]
> So in this case we could have serverless "generate" a LogGroup entry in the template and hope that API Gateway won't complain when it's trying to do the same, correct?

Yes, I believe that then we will gain control over that log group, through CF deployments. Still ofc that needs to be tested.

**Comment by pgrzesik** [comment]
@coyoteecd I believe that API Gateway first checks if such LogGroup exists and if that's not the case it tries to create one, that's how it works for regular logs so it's highly likely that it will work in a similar manner here.

**Comment by coyoteecd** [comment]
I can confirm that the above suggestion - i.e. generate a LogGroup as part of the stack deploying the API gateway - works fine. I have used this in a new project by defining the correct group as a custom resource that's being referenced from _serverless.ts_. For my needs, this was simple and "sufficiently clean" so I'm pasting it here in case others end up with the same question:

```yml
Resources:
  ApiGatewayExecutionLogGroup:
    Type: AWS::Logs::LogGroup
    Properties:
      LogGroupName: !Sub ['API-Gateway-Execution-Logs_${apiId}/${self:provider.stage}', { apiId: !Ref ApiGatewayRestApi }]
      RetentionInDays: ${self:provider.logRetentionInDays}
    UpdateReplacePolicy: Delete
    DeletionPolicy: Delete
```
